## Analysis of Player Purchasing in " Heroes of Pymoli"
by:  Andrew Guenthner
Date:  20 Mar 2019

Notes:  To run the analyses, the input file "purchase_data.csv" needs to be in a folder
called "Resources" in the same directory as this notebook.  

An accompanying "README" file should also be available in the same directory as this
notebook.  

In [1]:
# Import dependencies
import pandas as pd

# Set up file input
purchase_data_source = 'Resources/purchase_data.csv'
purchase_df = pd.read_csv(purchase_data_source)

In [2]:
# Check file header to make sure the DataFrame has loaded correctly
purchase_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

These data sets are expected to be clean and ready-to-go.  To verify, a simple check
for any missing values will suffice.

In [3]:
# Quick check for any missing values ...
purchase_df.isnull().sum()

Purchase ID    0
SN             0
Age            0
Gender         0
Item ID        0
Item Name      0
Price          0
dtype: int64

In [4]:
# Note the data types of each column for reference.
purchase_df.dtypes

Purchase ID      int64
SN              object
Age              int64
Gender          object
Item ID          int64
Item Name       object
Price          float64
dtype: object

### Initial exploratory analysis

Some basic characteristics of the data: 

1) Total number of players in the dataset (assuming one screen name per player):

In [5]:
# Compute total number of unique screen names
purchase_df.SN.nunique()

576

This is a reasonable size group for drawing inferences from the total set, but small enough that
we may need to be careful about analyzing subsets, such as particular age groups.

2) Total number of items purchased, the average transaction price, total number of transactions, and total revenue represented by these transactions.  For clarity, a "transaction" refers to the event of purchasing a single item recorded by the system, whereas a "purchase" refers to the player behavior of buying one or more items.  

In [6]:
# Compute total number of unique item codes
purchase_df['Item Name'].nunique()

179

In [7]:
# Compute average purchase price
print('${:.2f}'.format(purchase_df.Price.mean()))

$3.05


In [8]:
#Total number of transactions
purchase_df['Purchase ID'].count()

780

In [9]:
#Total revenue generated
print(f'${purchase_df.Price.sum():,.2f}')

$2,379.77


The transactions dataset is large enough to draw some meaningful conclusions, as long as deep sub-divisions are avoided.

### Gender Analysis

The breakdown of the data by gender is as follows:

1) Players by gender (based on unique screen name count)

In [10]:
# Make a general-purpose group-by item
purch_by_gender = purchase_df.groupby('Gender')
# Count unique screen names by group
gender_count = pd.DataFrame(purch_by_gender['SN'].nunique().reset_index())
# Generate % data and format the dataframe 
gender_count['% by Gender'] = 100 * gender_count['SN'] / gender_count['SN'].sum()
gender_count['% by Gender'] = gender_count['% by Gender'].map('{:.0f}'.format)
gender_count.style.hide_index()

The players identify as male by a large margin.

2) Purchase characteristics by gender, including number of transactions, average transaction amount, total purchases by each gender, and average spent on purchases per person by gender

In [11]:
# Count transactions by gender and start a new DataFrame
purch_count_gender = pd.DataFrame(purch_by_gender['Purchase ID'].count().reset_index())
# Average transaction amount by gender
price_avg_gender = pd.DataFrame(purch_by_gender['Price'].mean().reset_index())
# Total spend by gender
price_total_gender = pd.DataFrame(purch_by_gender['Price'].sum().reset_index())
# Merge these into a new dataframe for further analysis, and rename columns
merge_by_gender1 = purch_count_gender.merge(price_avg_gender,how='outer',on='Gender')
merge_by_gender2 = merge_by_gender1.merge(price_total_gender,how='outer',on='Gender')\
.rename(columns={'Purchase ID':'Transactions','Price_x':'Average Transaction','Price_y':'Total Spent'})
# Compute the average spent per player
merge_by_gender2['Average Spent per Player'] = merge_by_gender2['Total Spent'] / gender_count['SN']
# Format for display
merge_by_gender2['Average Transaction'] = merge_by_gender2['Average Transaction'].map('${:,.2f}'.format)
merge_by_gender2['Total Spent'] = merge_by_gender2['Total Spent'].map('${:,.2f}'.format)
merge_by_gender2['Average Spent per Player'] = merge_by_gender2['Average Spent per Player'].map('${:,.2f}'.format)
merge_by_gender2.style.hide_index()

Players identifying as male spent slightly less than those identifying as other genders, on both a per transaction and total per player basis.  However, because of the larger numbers, they represent by far the largest source of revenue.

Game players overwhelmingly identify as male.  The following table compares purchase behavior by gender.

### Age-Group Analysis

For age-based analysis, the purchase data was binned into five-year intervals, starting with under 10, 10-14, 15-19 ... ending with 40 and above (the maximum player age was 45).

In [12]:
# Set up bind and labels for age groups
bins = [0, 10, 15, 19, 24, 29, 34, 39, 46]
age_labels = ['Under 10','10-14', '15-19', '20-24', '25-29', '30-34', '35-39', '40 and over']
# Create age groups designation and append to DataFrame
purchase_df['Age Group'] = pd.cut(purchase_df.Age,bins,labels=age_labels)
# Create groupby with age groups 
purch_by_agegroup = purchase_df.groupby('Age Group')
# Start a new DataFrame with a re-labeled column head, showing count of items purchased by age group
purchase_byage_df = pd.DataFrame(purch_by_agegroup['Purchase ID'].count().reset_index().rename(columns={"Purchase ID":"Items Purchased"}))
# Make new columns with average price and total spend by age group
merge3 = purchase_byage_df.merge(pd.DataFrame(purch_by_agegroup['Price'].mean().rename('Average Item Price')), on='Age Group', how='outer')
purchase_byage_df = merge3.merge(pd.DataFrame(purch_by_agegroup['Price'].sum().rename('Total Revenue')), on='Age Group', how='outer')
# Make one new column with average spend per person
perperson_spend_by_age = purchase_df.groupby('Age Group').apply(lambda grp : grp.Price.sum() / grp.SN.nunique())
purchase_byage_df = purchase_byage_df.merge(pd.DataFrame(perperson_spend_by_age.rename('Revenue per Person')), on='Age Group', how='outer')
# Clean up the column formats for display 
purchase_byage_df['Average Item Price'] = purchase_byage_df['Average Item Price'].map('${:,.2f}'.format)
purchase_byage_df['Total Revenue'] = purchase_byage_df['Total Revenue'].map('${:,.2f}'.format)
purchase_byage_df['Revenue per Person'] = purchase_byage_df['Revenue per Person'].map('${:,.2f}'.format)
purchase_byage_df.style.hide_index()

Almost 50% of revenue was generated by 20-24 years olds, although on a per-person basis, their spending was not especially large.  On a per-person basis, players aged 35-39 outspent other age groups by about 20% per purchase, and players under 10 were also associated with a higher average item price and revenue per person.  

### Top purchasers, most popular items, and largest sources of revenue

The table below shows the key characteritics of the five players with the largest total purchase values.

In [13]:
# Generate the groupby DataFrame
top5_players = purchase_df.groupby('SN').agg({'Purchase ID':'count','Price':['mean','sum']}).reset_index()
# Sort and chop to top 5
top5_players = top5_players.sort_values([('Price','sum')],ascending=False).head(5)
# Prettify with new, single-level column names and formats
top5_players.columns = top5_players.columns.get_level_values(0)
top5_players.columns = ['SN','Items Purchased','Average Price','Total Spent']
top5_players['Average Price'] = top5_players['Average Price'].map('${:,.2f}'.format)
top5_players['Total Spent'] = top5_players['Total Spent'].map('${:,.2f}'.format)
# Display
top5_players.style.hide_index()

In this data set, the top 1% of spenders outspent the average player by a factor of roughly 3-4.   

The table below shows the key characteritics of the five most popular items.

In [14]:
# Generate the groupby DataFrame
top5_items = purchase_df.groupby('Item ID').agg({'Item Name':'first','Purchase ID':'count','Price':['mean','sum']}).reset_index()
# Sort and chop to top 5
top5_items = top5_items.sort_values([('Purchase ID','count')],ascending=False).head(5)
# Prettify with new, single-level column names and formats
top5_items.columns = top5_items.columns.get_level_values(0)
top5_items.columns = ['Item ID','Item Name','Times Purchased', 'Average Price','Total Spent']
top5_items['Average Price'] = top5_items['Average Price'].map('${:,.2f}'.format)
top5_items['Total Spent'] = top5_items['Total Spent'].map('${:,.2f}'.format)
# Display
top5_items.style.hide_index()

 In this data set, the top five items are purchased 2-3 x as often as the "average" (780 transactions for 179 items yields an average of somewhat over four purchases per item)

The table below shows the key characteritics of the five items that generated the most revenue.

In [15]:
# Generate the groupby DataFrame
top5_sales = purchase_df.groupby('Item ID').agg({'Item Name':'first','Purchase ID':'count','Price':['mean','sum']}).reset_index()
# Sort and chop to top 5
top5_sales = top5_sales.sort_values([('Price','sum')],ascending=False).head(5)
# Prettify with new, single-level column names and formats
top5_sales.columns = top5_sales.columns.get_level_values(0)
top5_sales.columns = ['Item ID','Item Name','Times Purchased', 'Average Price','Total Spent']
top5_sales['Average Price'] = top5_sales['Average Price'].map('${:,.2f}'.format)
top5_sales['Total Spent'] = top5_sales['Total Spent'].map('${:,.2f}'.format)
# Display
top5_sales.style.hide_index()

In this data set, the top five items by revenue generated 3-4 x the average revenue per item (\\$2,379.77 total revenue over 179 unique items yields an average of about \\$13.30 per item.